tokio-trace/tokio-trace-core/src/metadata.rs

```diff
@@ -0,0 +1,272 @@
+    /// Thus, these fields are *not* considered stable public API, and they may
+    /// change warning. Do not rely on any fields on `Metadata`. When
+    /// constructing new `Metadata`, use the `metadata!` macro or the
+    /// `Metadata::new` constructor instead!
+    #[doc(hidden)]
+    pub module_path: Option<&'a str>,
+
+    /// The name of the source code file where the span occurred, or `None` if
+    /// this could not be determined.
+    ///
+    /// **Warning**: The fields on this type are currently `pub` because it must
+    /// be able to be constructed statically by macros. However, when `const
+    /// fn`s are available on stable Rust, this will no longer be necessary.
+    /// Thus, these fields are *not* considered stable public API, and they may
+    /// change warning. Do not rely on any fields on `Metadata`. When
+    /// constructing new `Metadata`, use the `metadata!` macro or the
+    /// `Metadata::new` constructor instead!
+    #[doc(hidden)]
+    pub file: Option<&'a str>,
+
+    /// The line number in the source code file where the span occurred, or
+    /// `None` if this could not be determined.
+    ///
+    /// **Warning**: The fields on this type are currently `pub` because it must
+    /// be able to be constructed statically by macros. However, when `const
+    /// fn`s are available on stable Rust, this will no longer be necessary.
+    /// Thus, these fields are *not* considered stable public API, and they may
+    /// change warning. Do not rely on any fields on `Metadata`. When
+    /// constructing new `Metadata`, use the `metadata!` macro or the
+    /// `Metadata::new` constructor instead!
+    #[doc(hidden)]
+    pub line: Option<u32>,
+
+    /// The names of the key-value fields attached to the described span or
+    /// event.
+    ///
+    /// **Warning**: The fields on this type are currently `pub` because it must
+    /// be able to be constructed statically by macros. However, when `const
+    /// fn`s are available on stable Rust, this will no longer be necessary.
+    /// Thus, these fields are *not* considered stable public API, and they may
+    /// change warning. Do not rely on any fields on `Metadata`. When
+    /// constructing new `Metadata`, use the `metadata!` macro or the
+    /// `Metadata::new` constructor instead!
+    #[doc(hidden)]
+    pub fields: field::FieldSet,
+}
+
+/// Describes the level of verbosity of a `Span`.
+#[derive(Clone, Debug, Eq, PartialEq, Ord, PartialOrd)]
+pub struct Level(LevelInner);
+
+// ===== impl Metadata =====
+
+impl<'a> Metadata<'a> {
+    /// Construct new metadata for a span, with a name, target, level, field
+    /// names, and optional source code location.
+    pub fn new(
+        name: &'static str,
+        target: &'a str,
+        level: Level,
+        module_path: Option<&'a str>,
+        file: Option<&'a str>,
+        line: Option<u32>,
+        field_names: &'static [&'static str],
```

Comment:
`field_names` should not be exposed as `&[&str]`. We probably want an opaque type for this.

Reply:
In theory, I agree. In practice, since we can't use `const fn`s yet, the opaque type would need to have another constructor macro, so it just feels like a bunch of additional complexity to me. I can make that change regardless.

Reply:
(also, I think `FieldSet` is essentially the opaque type for this, but it also has a callsite ID. we can't construct that here since we need to create the metadata before the callsite)